graders/c/cgrader/cgrader.py

```diff
@@ -302,69 +301,90 @@ def test_run(
             name = "Test command: %s" % command
 
-        if exp_output is not None and not isinstance(exp_output, list):
+        if exp_output is None:
+            exp_output = []
+            must_match_all_outputs = True
+        elif not isinstance(exp_output, list):
             exp_output = [exp_output]
-        if reject_output is not None and not isinstance(reject_output, list):
+
+        if reject_output is None:
+            reject_output = []
+        elif not isinstance(reject_output, list):
             reject_output = [reject_output]
-        if msg is None and exp_output is not None:
-            msg = "Expected: %s" % (" AND " if must_match_all_outputs else " OR ").join(
-                [f"\n{t}\n" if "\n" in str(t) else f'"{t}"' for t in exp_output]
-            ) + (
-                ' but not "%s"' % '"/"'.join([str(t) for t in reject_output])
-                if reject_output
-                else ""
+
+        def compile_re(t):
+            if isinstance(t, re.Pattern):
+                return (t.pattern, t)
+            return (
+                str(t).strip(),
+                re.compile(
+                    "\\s+".join(map(re.escape, re.split("\\s+", str(t))))
+                    if ignore_consec_spaces
+                    else re.escape(str(t)),
+                    re.I if ignore_case else 0,
+                ),
             )
-        elif msg is None:
-            msg = ""
+
+        exp_output = [compile_re(t) for t in exp_output]
+        reject_output = [compile_re(t) for t in reject_output]
 
         out = self.run_command(
             command if args is None else ([command] + args),
             input,
             sandboxed=True,
             timeout=timeout,
         )
+
         outcmp = out
+        if highlight_matches and out:
+            for _, r in exp_output:
+                out = r.sub(f"\\033[32m\\g<0>\\033[0m", out)
+            for _, r in reject_output:
+                out = r.sub(f"\\033[31m\\g<0>\\033[0m", out)
         if not out:
             out = "(NO OUTPUT)"
         elif not out.endswith("\n"):
             out += "\n(NO ENDING LINE BREAK)"
 
-        if ignore_case:
-            outcmp = outcmp.lower()
-            if exp_output:
-                exp_output = [str(t).lower() for t in exp_output]
-            if reject_output:
-                reject_output = [str(t).lower() for t in reject_output]
-
-        if ignore_consec_spaces:
-            # Replace all space-like characters with single space
-            outcmp = re.sub(r"\s+", " ", outcmp)
-            if exp_output:
-                exp_output = [re.sub(r"\s+", " ", str(t)) for t in exp_output]
+        if msg is None and exp_output:
+            comment = (
+                ""
+                if len(exp_output) == 1
+                else " all of"
+                if must_match_all_outputs
+                else " one of"
+            )
+            msg = f"Expected{comment}:\n\t" + "\n\t".join(
```

Comment:
I don't think these need to be in a list (can delete the `[]` brackets) because join can take in any generator expression.

Reply:
Resolved in ea794457e.